File: LeetCodeRecord/MethodFirst/200-NumIslands.java
package MethodFirst;

import java.util.LinkedList;
import java.util.Queue;

class NumIslands {
    public int numIslands(char[][] grid) {
        int count = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    bfs(grid, i, j);
                    count++;
                }
            }
        }

        return count;
    }

    public void bfs(char[][] grid, int i, int j) {
        Queue<int[]> queue = new LinkedList<int[]>();
        queue.offer(new int[]{i, j});
        grid[i][j] = '0';

        while (!queue.isEmpty()) {
            int[] temp = queue.remove();
            int tempi = temp[0];
            int tempj = temp[1];

            if (tempi - 1 >= 0 && grid[tempi - 1][tempj] == '1') {
                grid[tempi - 1][tempj] = '0';
                queue.offer(new int[]{tempi - 1, tempj});
            }

            if (tempi + 1 < grid.length && grid[tempi + 1][tempj] == '1') {
                grid[tempi + 1][tempj] = '0';
                queue.offer(new int[]{tempi + 1, tempj});
            }
            
            if (tempj - 1 >= 0 && grid[tempi][tempj - 1] == '1') {
                grid[tempi][tempj - 1] = '0';
                queue.offer(new int[]{tempi, tempj - 1});
            }

            if (tempj + 1 < grid[0].length && grid[tempi][tempj + 1] == '1') {
                grid[tempi][tempj + 1] = '0';
                queue.offer(new int[]{tempi, tempj + 1});
            }
        }
    }
}

File: LeetCodeRecord/MethodFirst-Golang/200-NumIslands.go
package main

func numIslands(grid [][]byte) int {
	count := 0

	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			if grid[i][j] == '1' {
				dfs(grid, i, j)
				count++
			}
		}
	}

	return count
}

func dfs(grid [][]byte, i, j int) {
	if i < 0 || j < 0 || i >= len(grid) || j >= len(grid[0]) || grid[i][j] == '0' {
		return
	}

	grid[i][j] = '0'
	dfs(grid, i-1, j)
	dfs(grid, i+1, j)
	dfs(grid, i, j-1)
	dfs(grid, i, j+1)
}
